File: py-game/game/map_manager.py
from config import TILE_WIDTH, TILE_X_NUM, TILE_Y_NUM
from game.objects.man import ManObj
from game.objects.brick import BrickObj


class MapManager:
    """
    Manages the 2-D tile grid.
    Each cell is one of:
      None        -> empty
      "wall"      -> indestructible wall (string sentinel)
      BaseObj     -> any game object (brick, bomb, item)

    Players never occupy a map cell — their position is pixel-based.
    """

    def __init__(self, matrix: list[list], player_list: list[ManObj]):
        self.map: list[list] = []
        self._init_by_matrix(matrix, player_list)

    # ------------------------------------------------------------------ queries

    def is_tile_empty(self, ix: int, iy: int) -> bool:
        return self.map[iy][ix] is None

    def get_tile(self, ix: int, iy: int):
        return self.map[iy][ix]

    def get_tile_type(self, ix: int, iy: int) -> str:
        tile = self.map[iy][ix]
        if tile is None:
            return "empty"
        if tile == "wall":
            return "wall"
        return tile.get_obj_type()

    # ------------------------------------------------------------------ mutations

    def set_tile(self, ix: int, iy: int, obj):
        self.map[iy][ix] = obj

    def clear_tile(self, ix: int, iy: int):
        self.map[iy][ix] = None

    # ------------------------------------------------------------------ collision

    def can_man_move(self, pos_x: int, pos_y: int, man_obj: ManObj) -> bool:
        """
        Check all four corners of the man's tile-sized bounding box.
        Returns True if the move is legal.
        """
        corners = [
            (pos_x,                  pos_y),
            (pos_x + TILE_WIDTH - 1, pos_y),
            (pos_x,                  pos_y + TILE_WIDTH - 1),
            (pos_x + TILE_WIDTH - 1, pos_y + TILE_WIDTH - 1),
        ]
        for cx, cy in corners:
            ix = cx // TILE_WIDTH
            iy = cy // TILE_WIDTH
            tile_type = self.get_tile_type(ix, iy)

            if tile_type == "bomb":
                # Allow walking through a bomb if we were standing on it when placed
                can_pass = any(
                    bx <= cx < bx + TILE_WIDTH and by <= cy < by + TILE_WIDTH
                    for bx, by in man_obj.can_pass_bomb_pos_list
                )
                if not can_pass:
                    return False
            elif tile_type in ("brick", "wall"):
                return False

        return True

    # ------------------------------------------------------------------ init

    def _init_by_matrix(self, matrix: list[list], player_list: list[ManObj]):
        h = len(matrix)
        w = len(matrix[0])
        self.map = [[None] * w for _ in range(h)]

        for iy in range(h):
            for ix in range(w):
                cell = matrix[iy][ix]
                if cell == "wall":
                    self.map[iy][ix] = "wall"
                elif cell == "brick":
                    brick = BrickObj(ix, iy)
                    self.map[iy][ix] = brick
                elif cell in ("man1", "man2"):
                    man = ManObj(ix, iy, cell)
                    player_list.append(man)
                    self.map[iy][ix] = None   # players don't block tiles

File: backend/src/game/manager/map_manager.rs
use crate::game::config::TILE_WIDTH;
use crate::game::map_data::{MapData, MapMatrixCell};
use crate::game::manager::obj_manager::{MapMatrix, MapTile, MapTileType};
use crate::game::objects::brick::BrickObj;
use crate::game::objects::man::ManObj;
use crate::game::types::{ManSpriteKey, MapIndex, Position};
use crate::game::utils::tran_pos_to_index;

pub struct MapManager {
    pub map: MapMatrix,
}

// ========================
// Tile access helpers
// ========================
impl MapManager {
    pub fn get_tile_type(&self, index: &MapIndex) -> MapTileType {
        let (ix, iy) = (index.index_x as usize, index.index_y as usize);
        match self.map.get(iy).and_then(|row| row.get(ix)) {
            None | Some(None) => MapTileType::Empty,
            Some(Some(tile)) => tile.tile_type(),
        }
    }

    pub fn clean_tile(&mut self, index: &MapIndex) {
        let (ix, iy) = (index.index_x as usize, index.index_y as usize);
        if let Some(row) = self.map.get_mut(iy) {
            if let Some(cell) = row.get_mut(ix) {
                *cell = None;
            }
        }
    }

    pub fn set_tile(&mut self, index: &MapIndex, tile: MapTile) {
        let (ix, iy) = (index.index_x as usize, index.index_y as usize);
        if let Some(row) = self.map.get_mut(iy) {
            if let Some(cell) = row.get_mut(ix) {
                *cell = Some(tile);
            }
        }
    }

    pub fn take_tile(&mut self, index: &MapIndex) -> Option<MapTile> {
        let (ix, iy) = (index.index_x as usize, index.index_y as usize);
        self.map.get_mut(iy)?.get_mut(ix)?.take()
    }

    pub fn height(&self) -> u32 {
        self.map.len() as u32
    }

    pub fn width(&self) -> u32 {
        self.map.first().map_or(0, |row| row.len() as u32)
    }
}

impl MapManager {
    /// Build a MapManager from static map data and a list of participating players
    /// (ordered: first element gets MAN1 spawn, second gets MAN2, etc.).
    ///
    /// Returns the `MapManager` and the `ManObj` list populated from spawn cells.
    pub fn new(map_data: &MapData, player_man_keys: &[ManSpriteKey]) -> (Self, Vec<ManObj>) {
        let mut manager = Self { map: vec![] };
        let players = manager.init_by_map(map_data, player_man_keys);
        (manager, players)
    }

    // ========================
    // Movement validation
    // ========================

    pub fn can_man_move_by_position(&self, pos: &Position, man_obj: &ManObj) -> bool {
        let tw = TILE_WIDTH;
        let corners = [
            Position { pos_x: pos.pos_x, pos_y: pos.pos_y },
            Position { pos_x: pos.pos_x + tw - 1, pos_y: pos.pos_y },
            Position { pos_x: pos.pos_x, pos_y: pos.pos_y + tw - 1 },
            Position { pos_x: pos.pos_x + tw - 1, pos_y: pos.pos_y + tw - 1 },
        ];

        corners.iter().all(|p| {
            let map_index = tran_pos_to_index(p);
            let tile_type = self.get_tile_type(&map_index);

            match tile_type {
                MapTileType::Wall | MapTileType::Brick => false,
                MapTileType::Bomb => {
                    // Allow if the player is still overlapping the bomb tile they just placed
                    man_obj.can_pass_bomb_pos_list.iter().any(|bomb_pos| {
                        p.pos_x >= bomb_pos.pos_x
                            && p.pos_x < bomb_pos.pos_x + tw
                            && p.pos_y >= bomb_pos.pos_y
                            && p.pos_y < bomb_pos.pos_y + tw
                    })
                }
                _ => true,
            }
        })
    }

    // ========================
    // Initialisation
    // ========================

    fn init_by_map(
        &mut self,
        map_data: &MapData,
        player_man_keys: &[ManSpriteKey],
    ) -> Vec<ManObj> {
        let matrix = &map_data.matrix;
        let h = matrix.len();
        let w = matrix.first().map_or(0, |row| row.len());

        let mut map_matrix: MapMatrix = (0..h).map(|_| (0..w).map(|_| None).collect()).collect();
        let mut players: Vec<ManObj> = vec![];

        for (y, row) in matrix.iter().enumerate() {
            for (x, cell) in row.iter().enumerate() {
                let index = MapIndex {
                    index_x: x as u32,
                    index_y: y as u32,
                };

                match cell {
                    MapMatrixCell::Wall => {
                        map_matrix[y][x] = Some(MapTile::Wall);
                    }
                    MapMatrixCell::Brick => {
                        let brick = BrickObj::new(&index);
                        map_matrix[y][x] = Some(MapTile::Brick(brick));
                    }
                    MapMatrixCell::Empty => {
                        // leave as None
                    }
                    // Player spawn cells — create the ManObj at this position
                    spawn_cell => {
                        let man_key_opt = match spawn_cell {
                            MapMatrixCell::MAN1 => Some(ManSpriteKey::Man1),
                            MapMatrixCell::MAN2 => Some(ManSpriteKey::Man2),
                            MapMatrixCell::MAN3 => Some(ManSpriteKey::Man3),
                            MapMatrixCell::MAN4 => Some(ManSpriteKey::Man4),
                            _ => None,
                        };

                        if let Some(default_key) = man_key_opt {
                            // Use the caller-supplied key if provided (allows remapping slots).
                            let spawn_order = match default_key {
                                ManSpriteKey::Man1 => 0,
                                ManSpriteKey::Man2 => 1,
                                ManSpriteKey::Man3 => 2,
                                ManSpriteKey::Man4 => 3,
                            };
                            let man_key = player_man_keys
                                .get(spawn_order)
                                .cloned()
                                .unwrap_or(default_key);

                            let man = ManObj::new(&index, man_key);
                            players.push(man);
                            // Spawn cell is traversable — leave tile as None
                            map_matrix[y][x] = None;
                        }
                    }
                }
            }
        }

        self.map = map_matrix;
        players
    }
}
